### ISSUE TITLE
Create queue manager

### ISSUE THREAD

**Issue by cawarren** [Recidiviz/recidiviz-data#88]
_This is part of ingest generalization work_

**Background:**
To date, each scraper has been expected to get its own scraping (request throttling) queue. This was partly a matter of convenience while getting set-up, and partly because state and federal prison systems are large enough to keep a queue occupied around the clock.

As we look at how to scale this system to jails as well as prisons, this is no longer viable - there are ~3,000 county jails in the US, and each would take less than 5min to scrape.

As a result, we're switching to a new model. While some regions will continue to have special, dedicated queues (e.g., `us_ny`), others will lease queues which are pre-configured for the specific rate they need (e.g., 15 queries per minute). The list of acceptable queue / queue rates for a region will be listed in the `region_manifest` (see issue #84). Many regions may have multiple acceptable rates (e.g., 15/m and 20/m).

This requires a new part of the system, a `queue_manager` module.

**Expected workflow:**
The `queue_manager` is expected to work something like this:
- When scraper starts, `scraper_control` (`queue_control`, if issue #83 isn't complete yet) submits the region and requested lease period to the `queue_manager` queue (this may be a queue, or just a FIFO list on a datastore object - see _Scope of work_ below)
- At some interval (or when poked by `scraper_control`), `queue_manager` checks whether any of the queues it manages are free, and if so attempts to retrieve the next task for that queue type from its own `queue_manager` queue
- If there is such a task, it calls `scraper_control` with the region and new queue name that has been leased, which kicks off that scrape. `queue_manager` also creates a deferred task for itself, for the time of lease expiration, to wake back up and evict the scraper if it's still using that queue.
- When scraper finishes or stops, it calls `scraper_control.stop_scraper()` on itself, which calls `queue_manager` to terminate the lease and purge the queue (and check if others are waiting on it, to re-lease the queue).

_Note: GAE recommends not adding new tasks to a queue for up to 1m after purging it - use the [GAE deferred library](https://cloud.google.com/appengine/articles/deferred) to add delays without exceeding request deadlines._

For more details, see the design doc's [Life of a Scraper](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.qz0bdjml50ah) and [Queues](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.uwhtuqwgoo6) sections.

**Scope of work:**
- Evaluate whether to use a pull queue for scraping queue leases, rather than datastore lock entities. Update design doc with decision and any needed changes.
- Create the `queue_manager` module (see design doc's [Modules section](https://docs.google.com/document/d/11fv-hLwLOFcM5XSE16lYne5hnqFzcQK1vwvTBVVyhn8/edit#heading=h.v6780khuazyt))
- Build the workflow described above for within `queue_manager` module
- Update `scraper_control` to make use of the new `queue_manager`
- Add test coverage for the new component

**Comment by jessex** [comment]
After thinking about the implementation required to facilitate this design in our architecture, and the current state and pace of development, I'm thinking that maybe this work should be deferred. Here's my reasoning:
* The requirements to allow a region to wait for a queue to become available and for existing queues to be released in the event of completion or timeout or failure of the reserving ingest process, and the absolute necessity of ensuring we don't accidentally starve ourselves, makes this a pretty non-trivial bit of work to accomplish. It's a distributed scheduling problem, which are solved problems, but it's a bit beyond the scope of the GAE task queue system. It can be done, but not particularly simply and not without writing a bunch of logic a layer or two above where it belongs in the abstraction hierarchy (it's really too low level for app land). I have not yet found a library or framework that drops atop GAE task queues to make it "just work," so it's likely going to be by hand.
* The logical flow required looks something like the following, and there are certainly edge cases in here not accounted for. This is a good amount of business logic to write up, test, and get hardened operationally. It will take time and energy to get right.
```
1. scraper_control.scraper_start calls queue_manager.reserve_queue(scrape_key)
    1. Need to separate scraper_control.start logic into that which is invoked when we try to reserve a queue, and that which is invoked when we actually get a queue and wish to start
2. queue_manager determines set of acceptable queues based on region manifest
3. queue_manager checks if any of the acceptable queues are available (queue_manager.is_available(queue))
    1. Logic TBD…
4. If yes:
    1. queue_manager marks queue as reserved with max execution time
        1. Logic TBD…
    2. queue_manager initiates scraper.scraper_start
    3. queue_manager launches deferred task to periodically check on state of queue to see if it can be released
        1. Should use queue_manager.is_available(queue) to check if release is allowed
            1. If scraper has timed out, need to invoke scraper_control.scraper_stop
        2. If so, invokes queue_manager.release_queue(queue)
5. If no:
    1. queue_manager enqueues scrape_key as waiting for execution
        1. Logic TBD…
...
1. When scraper completes, or times out, or is directly stopped, or fails due to error, then a call is made to queue_manager.release_queue(queue)
    1. Check who is next in the scraper line and if they can use the queue that is available
    2. If yes: go to 4.1 and reserve it for the scraper
        1. Make sure to mark the scraper as no longer waiting
    3. If no: check the next waiting scraper for compatibility
    4. If no waiting scrapers can use the queue, mark it as released and return
```
* The reasons motivating these requirements are 1) that we want to allow ingest processes targeting specific source data systems to throttle to specific rates across all processes targeting that system simultaneously (e.g. snapshot and background scrapes against the same system, or one vendor's backend powering all of a state's county systems with the same servers), 2) that we want to allow different ingest processes targeting different source data systems to run without neighbor effects, and 3) that we want to avoid exhausting our resources, e.g. inefficient spending or bumping against the 100-queue cap for GAE projects (this can be raised through a manual request to GCP support in some cases).
* However, we are not anywhere near the 100-queue cap nor do we have a definite sense for how spending patterns will evolve with the system. But we can intuit that until we have hit a significant proportion of US corrections systems, a) the difference in costs between more processes running simultaneously on more queues versus more even distribution throughout the day will be small, b) the operational overhead of staggering start and stop times will be manageable, and c) we will have evolved architecturally anyway--for example, we could conceivably migrate to GCP Pub/Sub, which would handle this scheduling natively.

Given the amount of complexity this requires, our goals to rapidly onboard new systems and focus on scaling geopolitically, the fact that we won't see much benefit for such sophistication for some time, and that our architecture is evolving, I think we should, for now, just stand up a new queue with a desired throttle rate for each new system or group of systems we onboard. We can file a separate, simpler issue to insert into #90 in place of this one, and keep this one in the backlog for later.

Thoughts, @cawarren and @ohinds?

**Comment by jessex** [comment]
I've issued a PR for the above comment, assuming we agree: https://github.com/Recidiviz/recidiviz-data/pull/120

**Comment by jessex** [comment]
Sure enough, it looks like our leaves will indeed be easier: https://cloud.google.com/tasks/docs/queue-yaml

App Engine task queue is becoming Google Cloud Tasks, featuring a REST API (and client libraries) for all kinds of functionality, including queue management (see the above link). With the ability to create and destroy queues on the fly, we can switch from _statically defining queues with certain properties and then declaring which regions use which queues in their region manifests_ to _defining the queue properties each region needs in their manifests and then creating corresponding queues when it is time to start ingest, tearing them down when we stop_.

That should give us a much more scalable approach going forward. We won't have to worry about any sort of queue-of-queues or queue lock complexity, nor take the overall latency hit for doing so. We also don't have to keep adding to an increasingly unwieldy `queue.yaml` file as we expand. 

One bit of complexity that may remain, though, is where two regions share the same backing hardware and source system. If we wanted to throttle all ingests processes against that source system, across multiple regions, then we'd need to create an affinity between the regions so that we reuse the queue and keep it alive until all processes using it are stopped.

**Comment by jessex** [comment]
Okay, another remark after discussing this offline and thinking further:

**tl;dr:**
The new queue manager proposal is described in greater detail in the design doc. Once approved, we'll update the issue description accordingly.

**Part I:**
We deferred this work because the complexity described in my long comment on this PR, a few comments up. The ability to now create queues dynamically with Cloud Tasks API (beta) doesn't get rid of that complexity on its own. We still have the issue of handling ingest of regions which scrape from the same source data system and thus want to obey a shared rate limit. Broadly speaking, we can either ensure that such regions never run simultaneously, or we can bucket them in a combined rate limit.

The simplest approach here, which is now possible via the rest of the ingest generalization work (#90), is to just let those regions run on the same queue simultaneously. This should work fine and is described in greater depth in the ingest design doc. This can be done today by just including the same queue name in any two regions' manifests.

**Part II:**
However, the reason that the locking or queue-of-queues approaches were originally proposed is because App Engine has a hard limit on queues uploaded via queue.yaml: 100. If we are to scrape 3,000 county jails (it will definitely _not_ be 3,000 unique scrapers, but more than 100), then we would need to share queues with certain characteristics. However, _this_ is where Cloud Tasks API can help. Instead of hard-coding queues in queue.yaml and topping out at 100, we can create queues through the new API. I have been unable to find docs confirming whether or not there is a hard queue limit for locations or projects (to use their new terminology). If there is, we can destroy queues when they are no longer needed. If there isn't, we can forego that complexity and just keep queues around indefinitely since we know they will be continually reused.

**Comment by rasmi** [comment]
Speaking of deferring work, `deffered` is not yet available in the Cloud Tasks API, so I don't think it will be a swap-in migration as I was hoping. Pull queues are also not supported and are instead recommended to be implemented in Pub/Sub. See here: https://cloud.google.com/tasks/docs/migrating

What do you think, @jessex?

**Comment by jessex** [comment]
Rasmi's got the killer puns.

Yes we used `deferred` for that reason: to shovel async tasks off of the main thread, though we also also make use of the ability to schedule it for some point in the future, e.g. 30 seconds from now. Given that threads are supported now, I'd say we can swap in a package like [`concurrent.futures`,](https://docs.python.org/3/library/concurrent.futures.html) which is the standard library approach in Python 3.x (though that gives us a chicken and egg problem in that we need to upgrade to Cloud Tasks before upgrading to Python 3). But another Python 2-compatible package could be available, too, as an interim solution.

**Comment by jessex** [comment]
I made a note for myself to try to come up with an approach that could replace/augment the `docket` abstraction without use of pull queues, since I think it would be a shame to add in Cloud Pub/Sub just for that use case. But unless we have a great idea, I agree that that's probably the most straight-forward solution.

If you all come up with an idea before me, I'm all ears!

**Comment by jessex** [comment]
Some rough ideas for what we can do about the loss of pull queues with a migration to Cloud Tasks:
* Use Cloud Pub/Sub for the docket. It would be a very limited use for that service, which cuts both ways: it's a large dependency to add for a single purpose, but we can keep a tiny API surface for the docket and easily hide Pub/Sub as an implementation detail. Reading up on the object model and the delivery semantics, I think it would suit our docket needs just fine.
* Mix and match Cloud Tasks with GAE Task Queues. We'd use Cloud Tasks and its API for queue management for all of the push queues, but we'd keep using queue.yaml and the GAE SDK to use a single pull queue for the docket. This would not work in the second generation GAE runtime though, where those SDKs are no longer available. And obviously it wouldn't work if we switched from GAE as our runtime to, say, GKE.
* Switch the docket to use push queues. I haven't totally thought this through but I'm not sure why it wouldn't be possible. The docket as a pull queue lets us queue up work that we know we need to do for a background or snapshot scrape, often times (in background mode) with a single task that will then launch its own task tree. It lets an unacked task return to the docket to be retried if we don't ack it before some timeout. Its work is pulled down by a single worker handling a request. Its tasks are tagged with the region and scrape type so that the right scraper processes it. It holds items for all regions and scrape types which gives us a sort of natural ordering of all work, but that work gets grabbed from the docket at will and doesn't really serve any sort of global throttling purpose. A push queue should be able to accomplish all of the same, with the exception that unacked work cannot be automatically returned to the queue if, say, a worker just dies. But if a push queue-based task fails, then it will be retried automatically. I'd be curious what others think about this option, because if the push queue can serve the same purpose then it's the simplest route. +CC @cawarren who originally designed the docket and may know what I'm overlooking.

**Comment by rasmi** [comment]
@cawarren @jessex It's unclear to me why the docket necessary -- is it something to do with background scrapes vs. snapshot scrapes, request timeouts, or App Engine TaskQueue limitations? If we could exclusively use Cloud Tasks push queues, that would be great.

**Comment by jessex** [comment]
The docket is currently used to queue up tasks for snapshot scrapers: we read names out of persons out of Datastore, create a scrape task for each, and put it on the docket before starting a scrape task, which will take an item from the docket and launch more tasks in turn. Our new snapshot design should be able to get by only with push queues, though I know @terinpw is thinking about the exact mechanism for launching snapshot tasks now. I agree that it feels like a push queue should work for snapshotting without the need for a docket abstraction.

We also use the leasing functionality of the docket. If a task takes too long for any reason, then instead of potentially losing the task altogether then the item is returned back to the docket's pull queue to be grabbed by a future task that will try again. But it's arguable that this is unnecessary complexity for something that should almost never happen (the far, far more likely failure mode is a task failing outright and triggering retry instead of magically disappearing). If we trust push queue task failure to work, and we should, then the leasing functionality of the docket is not so necessary. 

The docket also lets us specifically pull tasks for region+scrape_type, i.e. if we are doing a background and snapshot scrape simultaneously then a worker can pull specifically for one or the other. This starts to get into the history of the scraper development so far, which began with NY which has tons of historical data and for whom the "background" scrape takes months to complete at our generously slow rate limit. That was what beget the snapshotting model and needs around being able to handle long-running tasks, hence sessions and the docket and such. I think that really long-running scrapes are not going to be the norm going forward. I think we should support them but the architecture should be built for the 99% case: scrapes which complete quickly.

Generally, the docket is also acting as a resting place for tasks that we know we need to process but may not be processing right now. It works in conjunction with the session abstraction to let us handle scrapes that may need to be paused and resumed. 

All of this is to say, I do think that we could go to push queues for the docket or some similar abstraction. Particularly if we get snapshotting working without the need for a pull queue and if we have a vision for how pause and resume can/should work without the need for a pull queue, then I think we can move forward with just push queues.

(Also, here's some archaeology for the adventurous: the PR (#46) that invented snapshot scraping and the docket. It has 290 comments spanning several weeks. The longer page-level comments about the docket itself and us learning about pull queues are the most interesting here. Keep in mind the context: we were only scraping the NY prison system at that point and its large historical data set, as opposed to being more optimized for smaller county jail systems that may or may not have history.)

**Comment by rasmi** [comment]
Thanks for the detailed response! To clarify -- my question was not about whether the docket should be push or pull, but why the docket exists at all. To me, the simplest architecture involves each region (or site, or collection of rate-limited sites) having its own push queue, with no docket involved. From what you've described, it seems like the functionality it provides is:

* Retries
  * Tasks now correspond to a single request, so if it times out, it's because the request failed, in which case we could rely on the push queue retry behavior. I think this is fine.
* A resting place for tasks which need to be process
  * Push queues can be "paused" to hold tasks -- is that sufficient, or do we need to be able to hold some tasks indefinitely while processing others regularly? I suppose we could create a paused push queue for some group of tasks as a holding queue and start it when ready.
* Simultaneous background + snapshot scrapes
  * It's unclear to me why this needs a docket -- is there an effective difference between snapshot scrapes and background scrapes in terms of implementation? I thought everything has switched to a "task", each task corresponding to one request, which wouldn't brush up against timeouts or other long-running task issues.
  * If we need to distinguish background vs. snapshot, could we have two push queues for each region, one for background and one for snapshot scrapes, that can be paused or started independently? (This may be convenient since queues can be created/destroyed programmatically).

**Comment by jessex** [comment]
To be more clear, my `tl;dr` from that comment was meant to agree that I think we can get rid of the docket abstraction, not just pull queues.

We actually do want background and snapshot work to use the same queue, because they'll be hitting the same website so they need to obey a shared rate limit. The docket was useful for snapshot scrapes because unlike in background scraping, where we would start at once place and then trace through a page tree as we go, snapshot scraping began by saying "here are the _n_ items to scrape now." So the request handler needed a place to put those n items where future tasks could grab them. But a push queue should suffice here, too, I think. That request handler (which may be replaced by another mechanism with the new snapshot model) would just added those items to the shared queue for that region and they'd be pushed out to workers at the configured rate limit.

So I think as regards migrating to Cloud Tasks, the progression is something like _remove the docket while maintaining existing scrape functionality_, and then in some order _migrate to Cloud Tasks_ and _build new snapshot process_, depending on what that new process calls for to make sure we don't build it by relying on the previous GAE Task Queue library.

Does that sound reasonable, @rasmi ?

**Comment by jessex** [comment]
>I'm not sure how common this scenario might be since jail populations tend to be quite small, but we should design for prisons too.

I agree completely.

>If my background scrape takes 5 days, each run through will add 5 days' worth of tasks to the queue, and the queue will never be emptied 

This is handled today via `sessions`. With each person we scrape, we update the open session for this region + scrape_type. When we hit the automated pause time, we make sure the session is updated and clear out anything residing in the queue. When we hit the automated resume time, we fetch that session and start again at the person that we last successfully scraped. This is because in the background scrape, we tend not to be dealing with a big list of docket tasks, but rather a starter task that is crawling over a page tree and creating new tasks as it goes. And eventually, a task discovers that it has reached the end of that page tree, and we close the session and walk away. The next time the background scrape is resumed, it sees from the session that we finished the full set of people last time and starts again from the top.

In this way, we can handle multi-day scrapes without a queue that is never emptied and consistently holding days worth of tasks.

**Comment by rasmi** [comment]
Some ideas after discussing with folks over here... It seems sessions are necessary in the case of a large system (like us_ny) whose search results may change in the amount of time it would take to scrape the entire history. For example, there are 1000 pages, and we can process 100 per day. By the time we get to 200, our "place"/index in the search system will have changed, and people may have been added before or after that place, so we would lose people along the way.

But I think this is also a matter of prioritizing the requests we're making. If we scraped only the roster first (not the person-details page), could we enqueue all the people in one night? Then for large background scrapes, we can have one queue for roster listings, and another for people details, and we can prioritize any items in the roster queue before any people details are scraped. Ideally this would allow us to avoid having to manage sessions across nights (assuming large sites can have their rosters collected quickly enough), but let me know if I'm missing something.

This also doesn't address the updating/snapshotting people in the db after a background scrape, which @terinpw is thinking about.

**Comment by jessex** [comment]
>By the time we get to 200, our "place"/index in the search system will have changed, and people may have been added before or after that place, so we would lose people along the way.

Yes this would definitely happen, and in the current logic we simply accept that there's an upper bound on latency between when a person first appears in such a system and when we first ingest their bookings. Of course, if a very large system really does take weeks or months to complete a single background scape, then that might be unacceptable latency.

>If we scraped only the roster first (not the person-details page), could we enqueue all the people in one night?

This is a good idea! I haven't thought through all of the implications but it appears useful for roster-based sites. But there are some sites, like NY itself, that do not have a single roster page to scrape, but rather many pages of smaller amounts of data. 

Some back of the envelope math: 305,000 people in NY DOCCS, you can get 4 people at a time with a "roster" request, it takes 1 request to get from the roster page to most person pages, but 2 requests to get to some person pages because of an intermediate disambiguation page. We performed around 13,000 requests per night for NY which got us a few thousand people ingested. Switching to this model, we'd get 52,000 person requests queued up in the same amount of time. So it would still take almost 6 days just to get the tasks for each person queued up.

Assuming we can make the shared rate limiting and prioritization among queues work, would it be possible to make it work even if we couldn't assemble an entire roster of person tasks in a single shot?

(Also, regardless of roster size, how do we handle failure that stops us short of assembling the person tasks, like the website going down or Cloud Tasks suffering some prolonged outage? I am guessing those are so exceptional that we are not thinking about addressing them straight away.)

**Comment by jessex** [comment]
@rasmi, @colincadams: do we have work lined up to take care of this issue after our migration to Cloud Tasks and to Pub/Sub? I think the spirit of the issue at this point should maybe be focused on getting us off of hard-coded queues from `queue.yaml` and `regions.yaml` and onto dynamically generated queues for a given region + scrape type.
